---
title: "Eicosanoids mediators of hypertension: gwas"
author: "Joonatan Palmu"
date: "`r format(Sys.time(), '%d.%m.%Y')`"
---


```{r options, include=FALSE}
knitr::opts_chunk$set(include = TRUE, echo = TRUE, message = FALSE,
                      results='asis', cache=FALSE, warning=FALSE)
options(knitr.kable.NA = "")
now <- format(Sys.time(), '%Y%m%d-%H%M%S')
```

```{r Remove temporary data, include = FALSE}
unlink("cache", recursive=TRUE)
dir.create("cache", showWarnings = FALSE)
dir.create("session", showWarnings = FALSE)
dir.create("report", showWarnings = FALSE)
```

# Libraries

<details>
  <summary>Open/Close</summary>

```{r libraries, results = 'hide'}
library(knitr)
library(tidyr)
library(broom)
library(readr)
library(dplyr)
library(purrr)
library(ggplot2)
library(parallel)
library(biomaRt)
require(qqplotr)
library(GWASTools)
```

</details>

# Functions

```{r functions}
snpmart <- useEnsembl(biomart="snp", dataset="hsapiens_snp")

getrsid <- function(snpmart, chr, pos) {
    lapply(seq(length(chr)), function(i) {
        query <- getBM(attributes=c("refsnp_id", "allele", "chrom_start", "chrom_strand"),
              filters = c("chr_name", "start", "end"),
              values = list(chr[i], pos[i], pos[i]),
              mart = snpmart) %>%
            pull(refsnp_id)
        ret <- ifelse(length(query) == 1, query[1], NA)
        message(sprintf("(%i/%i) chr : %s, pos : %s, id : %s", i, length(chr), chr[i], pos[i], ret))
        ret
    }) %>% map_chr(~ifelse(is.na(.x), NA, paste0(.x)))
}
```

# Variables for GWAS

```{r batch, eval = FALSE}
pheno.batch <- read_tsv("FR02_pheno.txt") %>%
    select(Sample_ID, BATCH)
```

```{r phenotype, eval  =FALSE}
pheno.riskscores <- readRDS("riskscores.rds") %>%
    rename_at(vars(contains("riskpersd")), funs(gsub("_riskpersd", "", .))) %>%
    rename_at(vars(contains("riskclass")), funs(gsub("_riskclass", "C", .)))
```

```{r mds, eval = FALSE}
mds_axes <- read_tsv("pca/merged_MDS_relateds_removed.tsv") %>% select(-X14, -SOL, -IID)
```

```{r covariates joined, eval = FALSE}
covariates <- left_join(pheno.riskscores, pheno.batch, by = "Sample_ID") %>%
    select(Sample_ID, BATCH, AGE, sex, fwdbonf, fwdaic, lasso, fwdbonfC) %>%
    mutate(fwdbonfC = case_when(fwdbonfC == 1 ~ 0,
                                fwdbonfC == 2 ~ 0,
                                fwdbonfC == 3 ~ 1,
                                fwdbonfC == 4 ~ 1)) %>%
           left_join(., mds_axes, by = c("Sample_ID" = "FID"))
```

# Sample files for GWAS

```{r export MR data, eval = FALSE}
lapply(seq(1, 22), function(x, df) {
    gwas.samples <- read_delim(sprintf("fr0207_METAG_chr%s.samples", x), " ")
    joined <- gwas.samples %>%
        filter(ID_1 != 0) %>%
        left_join(., df, by = c("ID_1" = "Sample_ID")) %>%
        tibble::add_row(ID_1 = "0",
                        ID_2 = "0",
                        missing = "0",
                        fwdbonf = "P",
                        fwdbonfC = "B",
                        fwdaic = "P",
                        BATCH = "D",
                        lasso = "P",
                        AGE = "C",
                        sex = "D",
                        C1 ="C",
                        C2 = "C",
                        C3 = "C",
                        C4 = "C",
                        C5 = "C",
                        C6 = "C",
                        C7 = "C",
                        C8 = "C",
                        C9 = "C",
                        C10 = "C",
                        .before = 1) 
    write_delim(joined, sprintf("samples/riskscore_chr%s.samples", x), " ")
    excluded <- joined %>%
        filter(is.na(fwdbonf), ID_1 != 0) %>%
        pull(ID_1) %>%
        as.data.frame
    write_delim(excluded, sprintf("samples/exclusions_chr%s.exclude", x), " ", col_names = FALSE)
    TRUE
}, df = covariates)
```

For running the snptest

```
for n in $(seq 1 22); do grun.py -n "gwas-$n" -c "palmu-snptest.sh $n" -q havulinna.q --log-dir logs/; done
```

## Read gwas results

```{r gwas results}
results <- lapply(seq(1, 22), function(x) {
    read_delim(sprintf("results/output_%s.out", x), " ", skip = 13)
})

results.combined <- results %>%
    map_df(~as.data.frame(.x)) %>%
    mutate(chr = as.numeric(gsub("chr(.*):.*", "\\1", rsid)),
           start = as.numeric(gsub(".*:([^_]*)_.*", "\\1", rsid))) %>%
    filter(!is.na(start), !is.na(chr)) %>%
    filter(cohort_1_hwe > 1e-10,
           all_maf > 0.01,
           nchar(alleleA) == 1,
           nchar(alleleB) == 1) 
```

# Manhattan plot

```{r manhattan plot}
don <- results.combined %>% 
    group_by(chr) %>% 
    summarise(chr_len = max(start)) %>% 
    mutate(cumpos = cumsum(chr_len) - chr_len) %>%
    dplyr::select(-chr_len)  %>%
    left_join(results.combined, ., by=c("chr" = "chr")) %>%
    mutate(startcum = start + cumpos) %>%
    arrange(chr, start)

axisdf <- don %>%
    group_by(chr) %>%
    summarize(center = (max(as.numeric(startcum)) + min(as.numeric(startcum)))/2)

g.manhattan <- ggplot(don, aes(x = startcum, y = -log10(frequentist_add_pvalue))) +
    geom_point(aes(color=as.factor(chr)), alpha=0.8, size=0.2) +
    geom_hline(yintercept = -log10(5e-8), linetype="dashed") +
    scale_color_manual(values = rep(c("grey", "skyblue"), 22)) +
    scale_x_continuous(label = axisdf$chr, breaks = axisdf$center) +
    ylab("-log10(P)") +
    scale_y_continuous(expand = c(0, 0)) +
    theme_bw() +
    theme(axis.title.x=element_blank(),
          legend.position="none",
          panel.border = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank())

ggsave("cache/manhattan.png", g.manhattan, width = 10, height = 4, unit = "in")
```

<img src = "cache/manhattan.png" >

# QQ-plot

```{r qqplot}
png("cache/qqplot.png", 1000, 1000)
qqPlot(results.combined %>%
       dplyr::filter(!(chr == 11 & frequentist_add_pvalue < 5e-8)) %>%
       pull(frequentist_add_pvalue))
dev.off()
```

# Exporting results for MR

```{r test}
results.formrbase <- results.combined %>%
    filter(frequentist_add_pvalue < 5e-8,
           cohort_1_hwe > 1e-6,
           all_maf > 0.01,
           info  > 0.4,
           nchar(alleleA) == 1,
           nchar(alleleB) == 1) %>% 
    dplyr::select(rsid,
                  alleleA,
                  alleleB,
                  all_maf,
                  frequentist_add_info,
                  frequentist_add_beta_1,
                  frequentist_add_se_1,
                  frequentist_add_pvalue,
                  chr,
                  start) %>%
    dplyr::mutate(snip = getrsid(snpmart, chr, start))

write_csv(results.formrbase %>% filter(!is.na(snip)), path = "eicgwas.csv")
```

```{r top snips}
results.formrbase %>%
    group_by(chr) %>%
    arrange(frequentist_add_pvalue) %>%
    top_n(2) %>%
    dplyr::select(snip, chr, start, frequentist_add_pvalue)
```


```{r save session}
save.image(file = paste0("session/session-", now, ".Rdata"))
```
